### ISSUE TITLE
Symbol's function definition is void: -let*

### ISSUE THREAD

**Issue by dyba** [Fuco1/smartparens#420]
I'm giving smartparens a try, but I'm unable to get past this error message:

```
Symbol's function definition is void: -let*
```

This happens immediately after I require the config file in my ```init.el```. I ran ```ack``` on my ```.emacs.d``` folder and found [this line](https://github.com/Fuco1/smartparens/blob/98d725981a1eb12c14ac38447bde3830755b9e32/smartparens.el#L3164) in ```smartparens.el```:

``` emacs-lisp
(defun sp-insert-pair (&optional pair)
  "Automatically insert the closing pair if it is allowed in current context.
If PAIR is provided, use this as pair ID instead of looking
through the recent history of pressed keys.
You can disable this feature completely for all modes and all pairs by
setting `sp-autoinsert-pair' to nil.
You can globally disable insertion of closing pair if point is
followed by the matching opening pair.  It is disabled by
default.  See `sp-autoinsert-if-followed-by-same' for more info."
  (-let* ((active-pair (sp--pair-to-insert))
          ((open-pair close-pair trig) (sp--insert-pair-get-pair-info active-pair)))
    ;; Test "repeat last wrap" here.  If we wrap a region and then
    ;; type in a pair, wrap again around the last active region.  This
    ;; should probably be tested in the `self-insert-command'
    ;; advice... but we're lazy :D
    (if (and sp-autowrap-region ...
```

I've not written anything in Emacs Lisp before, so I don't know why that symbol is void. Is that an emacs built-in function? Do I need to download another library to have that function defined?

Thanks for your help in advance and for the work you are doing!

**Comment by dyba** [comment]
I don't know if this helps but I'm adding my list of dependencies and my ```init.el``` file:

```
ac-cider-20141231.901
archives
async-20141229.2221
auto-complete-20141228.633
caml-20130904.812
cider-20150112.840
clojure-cheatsheet-20140316.1520
clojure-mode-20150111.952
clojure-mode-20150113.1048
clojurescript-mode-0.5
coffee-mode-20140713.1902
color-theme-20080305.834
color-theme-sanityinc-solarized-20140619.659
color-theme-sanityinc-solarized-readme.txt
color-theme-sanityinc-solarized-readme.txt~
creole-20140705.414
dash-20140627.2341
db-20140421.1411
dockerfile-mode-20141208.1939
elixir-mode-20140705.1500
elnode-20140203.1506
epl-20140620.218
esxml-20140320.1321
fakir-20140417.1347
feature-mode-20140722.1629
git-commit-mode-20140605.520
git-rebase-mode-20140605.520
haskell-mode-20141104.534
helm-20150114.140
highlight-parentheses-20140620.1258
highlight-parentheses-readme.txt
kv-20140108.734
magit-20140623.1208
noflet-20141102.654
noflet-readme.txt
org-trello-20140608.1036
org-trello-readme.txt
org-trello-readme.txt~
paredit-20141206.954
pkg-info-20140610.630
popup-20141215.349
projectile-20140706.126
queue-0.1.1
racket-mode-20141023.637
rainbow-delimiters-20140329.934
request-20140316.417
s-20140620.1657
smartparens-20150112.902
solarized-theme-20140616.401
sunrise-commander-6.450
tuareg-20140707.1235
web-20140612.1643
```

```emacs-lisp
(require 'package)

(add-to-list 'package-archives
	     '("marmalade" . "http://marmalade-repo.org/packages/"))
(add-to-list 'package-archives
             '("melpa" . "http://melpa.milkbox.net/packages/") t)
(package-initialize)

;; Grab the following packages if not already installed
(defvar my-packages
  '(pkg-info
    dash
    auto-complete
    ac-cider
    cider
    projectile
    eldoc
    paredit
    clojure-mode
    rainbow-delimiters
    color-theme-sanityinc-solarized
    tuareg
    dockerfile-mode
    racket-mode
    haskell-mode
    feature-mode
    coffee-mode
    elixir-mode))

(dolist (p my-packages)
  (unless (package-installed-p p)
    (package-refresh-contents)
    (package-install p)))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Org-Mode Configuration
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(global-set-key "\C-cl" 'org-store-link)
(global-set-key "\C-cc" 'org-capture)
(global-set-key "\C-ca" 'org-agenda)
(global-set-key "\C-cb" 'org-iswitchb)

(setq org-log-done 'time)

(global-linum-mode)

;; Enable projectile globally
(require 'projectile)
(projectile-global-mode)

;; Highlight Parentheses
(require 'paren)
(setq show-paren-style 'parenthesis)
(show-paren-mode +1)

;; Load the path in which you store rbenv
;; (setenv "PATH" (concat "/usr/local/bin" ":" (getenv "PATH")))

;;(setq line-number-display-limit-width 15)

;; Remove the default splash screen
(setq inhibit-startup-message t)
;; Don't add any comments to the *scratch* buffer
(setq initial-scratch-message nil)

;; Need the following to have the sass executable available for use
;; with scss mode
(setq scss-sass-command "~/bin/sass")

;; Require text files end in a newline
(setq require-final-newline 't)

;; Autoload the scss-mode file
(add-to-list 'load-path (expand-file-name "~/.emacs.d/modes/scss-mode"))
(autoload 'scss-mode "scss-mode")
(add-to-list 'auto-mode-alist '("\\.scss\\'" . scss-mode))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Elixir Configuration
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(require 'elixir-mode)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Erlang Configuration
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(add-to-list 'load-path "/usr/local/opt/erlang/lib/erlang/lib/tools-2.7.1/emacs")

(add-to-list 'auto-mode-alist '("\\.erl?$" . erlang-mode))
(add-to-list 'auto-mode-alist '("\\.hrl?$" . erlang-mode))

(setq erlang-root-dir "/usr/local/opt/erlang/lib/erlang")
(add-to-list 'exec-path "/usr/local/opt/erlang/lib/erlang/bin")
(setq erlang-man-root-dir "/usr/local/opt/erlang/lib/erlang/man")

(require 'erlang-start)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Haskell Mode Configuration
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(require 'inf-haskell)

(add-hook 'haskell-mode-hook 'turn-on-haskell-doc-mode)
(add-hook 'haskell-mode-hook 'turn-on-haskell-indent)
(add-hook 'haskell-mode-hook 'interactive-haskell-mode)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Coq Mode Configuration
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(add-to-list 'load-path "/usr/local/opt/coq/lib/emacs/site-lisp")

(setq auto-mode-alist (cons '("\\.v$" . coq-mode) auto-mode-alist))
(autoload 'coq-mode "coq" "Major mode for editing Coq vernacular." t)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; ProofGeneral Configuration
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

;; (load-file "~/.emacs.d/vendor/ProofGeneral/generic/proof-site.el")

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Tuareg Mode Configuration (For OCaml)
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(require 'tuareg)

(setq auto-mode-alist 
      (append '(("\\.ml[ily]?$" . tuareg-mode))
	      auto-mode-alist))

;; -- Tweaks for OS X -------------------------------------
;; Tweak for problem on OS X where Emacs.app doesn't run the right
;; init scripts when invoking a sub-shell
(cond
 ((eq window-system 'ns) ; macosx
  ;; Invoke login shells, so that .profile or .bash_profile is read
  (setq shell-command-switch "-lc")))

;; -- opam and utop setup --------------------------------
;; Setup environment variables using opam
;; The following simpler alternative works as of opam 1.1
(dolist
   (var (car (read-from-string
	      (shell-command-to-string "opam config env --sexp"))))
  (setenv (car var) (cadr var)))
;; Update the emacs path
(setq exec-path (split-string (getenv "PATH") path-separator))
;; Update the emacs load path
(push (concat (getenv "OCAML_TOPLEVEL_PATH")
	      "/../../share/emacs/site-lisp") load-path)
;; Automatically load utop.el
(autoload 'utop "utop" "Toplevel for OCaml" t)
(autoload 'utop-setup-ocaml-buffer "utop" "Toplevel for OCaml" t)
(add-hook 'tuareg-mode-hook 'utop-setup-ocaml-buffer)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Clojure Mode Configuration
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(require 'clojure-mode)

;; For better indentation for Compojure macros
(define-clojure-indent
  (defroutes 'defun)
  (GET 2)
  (POST 2)
  (PUT 2)
  (DELETE 2)
  (HEAD 2)
  (ANY 2)

  ;; For better indentation of speclj
  (it 'defun)
  (describe 'defun)
  (context 'defun))

;; 'Always 2 spaces' indentation
(setq clojure-defun-style-default-indent t)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Cider Configuration
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

;; Enable ElDoc in Clojure buffers
(add-hook 'cider-mode-hook #'eldoc-mode)
(add-hook 'cider-repl-mode-hook #'paredit-mode)
(add-hook 'cider-repl-mode-hook #'rainbow-delimiters-mode)

;; log communication with the nREPL server
(setq nrepl-log-messages)

;; hide *nrepl-connection* and *nrepl-server* buffers from C-x b
(setq nrepl-hide-special-buffers t)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; SML Configuration
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

;; Add the SML/NJ directory to the Emacs path
(setenv "PATH" (concat "/usr/local/Cellar/smlnj/110.75/libexec/bin:"
                       (getenv "PATH")))
(setq exec-path (cons "/usr/local/Cellar/smlnj/110.75/libexec/bin" exec-path))

;; Have Emacs automatically place the buffer in SML mode whenever it sees
;; a file containing the .sml extension
(add-to-list 'auto-mode-alist '("\\.\\(sml\\|sig\\)\\'" . sml-mode))

(defun sml-mode-hook-defaults ()
  "Global defaults for SML mode"
  (setq sml-indent-level 4) ; indent level should be 4
  (setq indent-tabs-mode nil) ; never indent with tabs
  ; Bind the key C-c C-s to M-x sml-run
  (define-key sml-run "\C-c\C-s" 'sml-run))

(defun inferior-sml-mode-hook-defaults ()
  "Global defaults for the SML buffer"
  (define-key inferior-sml-mode-map "\C-cd" 'sml-prog-proc-chdir))

(add-hook 'sml-mode-hook 'sml-mode-hook-defaults)

(defun inf-sml-mode-hook-defaults ()
  "Local defaults for inferior SML mode"
  (define-key inferior-sml-mode-map "\C-cd" 'sml-cd))

;; Set defaults when loading SML files
(add-hook 'sml-mode-hook 'sml-mode-hook-defaults)
(add-hook 'inferior-sml-mode-hook 'inferior-sml-mode-hook-defaults)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; ParEdit Configuration
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

;; Turning on Paredit mode
(defun enable-paredit () (paredit-mode t))

;; Paredit hooks
(add-hook 'clojure-mode-hook 'enable-paredit)
(add-hook 'emacs-lisp-mode-hook 'enable-paredit)
(add-hook 'eval-expression-minibuffer-setup-hook 'enable-paredit)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; ElDoc Configuration
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

;; Requiring ElDoc
(require 'eldoc)

;; Make ElDoc aware of ParEdit's most used commands
(eldoc-add-command
 'paredit-backward-delete
 'paredit-close-round)

(defun enable-eldoc-defaults ()
  (turn-on-eldoc-mode)
  (eldoc-add-command
   'paredit-backward-delete
   'pradedit-close-round))

;; Adding Electric Return
(defvar electrify-return-match
  "[\]}\)\"]"
  "If this regexp matches the text after the cursor, do an \"electric\" return.")

(defun electrify-return-if-match (arg)
  "If the text after the cursor matches `electrify-return-match` then open and indent an empty line between the cursor and the text. Move the cursor to the new line."
  (interactive "P")
  (let ((case-fold-search nil))
    (if (looking-at electrify-return-match)
	(save-excursion (newline-and-indent)))
    (newline arg)
    (indent-according-to-mode)))

(defun enable-lisp-defaults ()
  (paredit-mode t)
  (turn-on-eldoc-mode)
  (eldoc-add-command
   'paredit-backward-delete
   'paredit-close-round)
  (local-set-key (kbd "RET") 'electrify-return-if-match)
  (eldoc-add-command 'electrify-return-if-match)
  (show-paren-mode t))

(add-hook 'emacs-lisp-mode-hook 'enable-lisp-defaults)
(add-hook 'clojure-mode-hook 'enable-lisp-defaults)

(set-frame-font "Menlo-16")
(custom-set-variables
 ;; custom-set-variables was added by Custom.
 ;; If you edit it by hand, you could mess it up, so be careful.
 ;; Your init file should contain only one such instance.
 ;; If there is more than one, they won't work right.
 '(ansi-color-faces-vector [default bold shadow italic underline bold bold-italic bold])
 '(ansi-color-names-vector (vector "#708183" "#c60007" "#728a05" "#a57705" "#2075c7" "#c61b6e" "#259185" "#042028"))
 '(custom-enabled-themes (quote (sanityinc-solarized-light)))
 '(custom-safe-themes (quote ("4cf3221feff536e2b3385209e9b9dc4c2e0818a69a1cdb4b522756bcdf4e00a4" "4aee8551b53a43a883cb0b7f3255d6859d766b6c5e14bcb01bed572fcbef4328" "1e7e097ec8cb1f8c3a912d7e1e0331caeed49fef6cff220be63bd2a6ba4cc365" "fc5fcb6f1f1c1bc01305694c59a1a861b008c534cae8d0e48e4d5e81ad718bc6" default)))
 '(fci-rule-color "#0a2832")
 '(vc-annotate-background nil)
 '(vc-annotate-color-map (quote ((20 . "#c60007") (40 . "#bd3612") (60 . "#a57705") (80 . "#728a05") (100 . "#259185") (120 . "#2075c7") (140 . "#c61b6e") (160 . "#5859b7") (180 . "#c60007") (200 . "#bd3612") (220 . "#a57705") (240 . "#728a05") (260 . "#259185") (280 . "#2075c7") (300 . "#c61b6e") (320 . "#5859b7") (340 . "#c60007") (360 . "#bd3612"))))
 '(vc-annotate-very-old-color nil))
(custom-set-faces
 ;; custom-set-faces was added by Custom.
 ;; If you edit it by hand, you could mess it up, so be careful.
 ;; Your init file should contain only one such instance.
 ;; If there is more than one, they won't work right.
 )
```

**Comment by Fuco1** [comment]
This is what happens to you when you use MELPA and aren't aware of it's "dangers". Smartparens properly declares its dependance on `dash` 2.10 where `-let*` comes from. But MELPA mangles the version numbers, giving you `dash` 20140600.something which is always more than 2.10. This way, your dependencies are *automatically always met* which then causes these errors.

When you install new package and use MELPA, you *must* always do an update of its dependencies manually.  This is exactly the antithesis of what a packaga system should do, but then, the authors of package.el never could've thought someone would puposefully fuck up the version numbers :/

So **TD;DR**: update dash.el package.

**Comment by dyba** [comment]
@Fuco1 Thank you for that explanation! I've heard about a stable version of MELPA. Do you recommend using that instead of regular MELPA? In the mean time I'll update the dash package.

**Comment by Fuco1** [comment]
MELPA Stable is a bit better, but as soon as you pull anything from the unstable one, all packages start to go through that (rather, if you even have it in your archive list).

You can fix it with repo pinning in 24.4, but it's bothersome and requires manual attention.

In the 24.5 master there's now a patch from @jorgenschaefer where you can set repo priority.  Hopefully the next release will come soon.

I'm closing as the issue is resolved, but if you have some more questions, feel free to ask here.
